xskillscore/core/deterministic.py

```diff
@@ -1,35 +1,86 @@
     axis = tuple(range(-1, -len(dim) - 1, -1))
     return dim, axis
 
 
-def pearson_r(a, b, dim):
+def _preprocess_weights(a, dim, new_dim, weights):
+    """Preprocesses weights array to prepare for numpy computation.
+
+    Parameters
+    ----------
+    a : xarray.Dataset or xarray.DataArray
+        One of the arrays over which the function will be applied.
+    dim : str, list
+        The original dimension(s) to apply the function along.
+    new_dim : str
+        The newly named dimension after running ``_preprocess_dims``
+    weights : xarray.Dataset or xarray.DataArray
+        Weights to apply to function, matching the dimension size of
+        ``new_dim``.
+    """
+    if weights is None:
+        return xr.full_like(a, None)  # Return nan weighting array.
+    else:
+        # Scale weights to vary from 0 to 1.
+        weights = weights / weights.max()
+        # Throw error if there are negative weights.
+        if weights.min() < 0:
+            raise ValueError(
+                "Weights has a minimum below 0. Please submit a weights array "
+                "of positive numbers."
+            )
+        # Check that the weights array has the same size
+        # dimension(s) as those being applied over.
+        drop_dims = [i for i in a.dims if i not in new_dim]
+        drop_dims = {k: 0 for k in drop_dims}
+        if dict(weights.sizes) != dict(a.isel(drop_dims).sizes):
+            raise ValueError(
+                f"weights dimension(s) {dim} of size {dict(weights.sizes)} "
+                f"does not match DataArray's size "
+                f"{dict(a.isel(drop_dims).sizes)}"
+            )
+        if dict(weights.sizes) != dict(a.sizes):
+            # Broadcast weights to full size of main object.
+            _, weights = xr.broadcast(a, weights)
+        return weights
+
+
+def pearson_r(a, b, dim, weights=None):
     """
     Pearson's correlation coefficient.
 
     Parameters
     ----------
-    a : Dataset, DataArray, GroupBy, Variable, numpy/dask arrays or scalars
-        Mix of labeled and/or unlabeled arrays to which to apply the function.
-    b : Dataset, DataArray, GroupBy, Variable, numpy/dask arrays or scalars
-        Mix of labeled and/or unlabeled arrays to which to apply the function.
+    a : xarray.Dataset or xarray.DataArray
+        Labeled array(s) over which to apply the function.
+    b : xarray.Dataset or xarray.DataArray
+        Labeled array(s) over which to apply the function.
     dim : str, list
         The dimension(s) to apply the correlation along.
+    weights : xarray.Dataset or xarray.DataArray
+        Weights matching dimensions of ``dim`` to apply during the function.
+        If None, an array of ones will be applied (i.e., no weighting).
 
     Returns
     -------
-    Single value or tuple of Dataset, DataArray, Variable, dask.array.Array or
-    numpy.ndarray, the first type on that list to appear on an input.
+    xr.DataArray or xr.Dataset
```

Comment:
Since the Parameters use the non-abbrieviated naming convention, maybe the Returns should too? And xr.Dataset comes first
```
    a : xarray.Dataset or xarray.DataArray
        Labeled array(s) over which to apply the function.
    b : xarray.Dataset or xarray.DataArray
        Labeled array(s) over which to apply the function.
```

Reply:
Oops :) too late. Doesn't really matter.